efibootmanager_policy.go

```diff
@@ -0,0 +1,335 @@
+		}
+		h.Write(b)
+
+		// 6) Get the Attribute Certificate Table address and size from the Certificate Table entry.
+		certTable = &dd[certTableIndex]
+	}
+
+	// 7) Exclude the Certificate Table entry from the calculation and hash	everything from the end of the Certificate Table entry
+	// to the end of image header, including the Section Table. The Certificate Table entry is 8 bytes long.
+	if certTable != nil {
+		hr.Seek(8, io.SeekCurrent)
+	}
+	b, err = ioutil.ReadAll(hr)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot read remainder of headers and section table: %w", err)
+	}
+	h.Write(b)
+
+	// 8) Create a counter called sumOfBytesHashed, which is not part of the signature. Set this counter to the SizeOfHeaders field.
+	sumOfBytesHashed := sizeOfHeaders
+
+	// 9) Build a temporary table of pointers to all of the section headers in the image. Do not include any section headers in the
+	// table whose Size field is zero.
+	var sections []*pe.SectionHeader
+	for _, section := range p.Sections {
+		if section.Size == 0 {
+			continue
+		}
+		sections = append(sections, &section.SectionHeader)
+	}
+
+	// 10) Using the Offset field in the referenced SectionHeader structure as a key, arrange the table's elements in ascending order.
+	// In other words, sort the section headers in ascending order according to the disk-file offset of the sections.
+	sort.Slice(sections, func(i, j int) bool { return sections[i].Offset < sections[j].Offset })
+
+	for _, section := range sections {
+		// 11) Walk through the sorted table, load the corresponding section into memory, and hash the entire section. Use the
+		// Size field in the SectionHeader structure to determine the amount of data to hash.
+		sr := io.NewSectionReader(r, int64(section.Offset), int64(section.Size))
+		b, err := ioutil.ReadAll(sr)
+		if err != nil {
+			return nil, xerrors.Errorf("cannot read section %s: %w", section.Name, err)
+		}
+		h.Write(b)
+
+		// 12) Add the section’s Size value to sumOfBytesHashed.
+		sumOfBytesHashed += int64(section.Size)
+
+		// 13) Repeat steps 11 and 12 for all of the sections in the sorted table.
+	}
+
+	// 14) Create a value called fileSize, which is not part of the signature. Set this value to the image’s file size. If fileSize is
+	// greater than sumOfBytesHashed, the file contains extra data that must be added to the hash. This data begins at the
+	// sumOfBytesHashed file offset, and its length is:
+	// fileSize – (certTable.Size + sumOfBytesHashed)
+	fileSize, err := r.Size()
+	if err != nil {
+		return nil, xerrors.Errorf("cannot obtain image size: %w", err)
+	}
+
+	if fileSize > sumOfBytesHashed {
+		var certSize int64
+		if certTable != nil {
+			certSize = int64(certTable.Size)
+		}
+
+		if fileSize < (sumOfBytesHashed + certSize) {
+			return nil, errors.New("image too short")
+		}
+
+		b = make([]byte, fileSize-sumOfBytesHashed-certSize)
```

Comment:
same question about reasonable size?

Reply:
I've add a helper to hash data 1 page at a time now